code/controllers/subsystem/garbage.dm

```diff
@@ -64,97 +70,178 @@ SUBSYSTEM_DEF(garbage)
+	var/static/lastlevel
+	var/static/list/queued
+	if (queued) //runtime last run before we could do this.
+		queues[lastlevel] -= queued
+
+	queued = list()
+	lastlevel = level
+
+	for (var/refID in queue)
 		if (MC_TICK_CHECK)
 			break
-		var/refID = queue[1]
+
+		queued += refID
+
 		if (!refID)
-			queue.Cut(1, 2)
 			continue
 
 		var/GCd_at_time = queue[refID]
-		if(GCd_at_time > time_to_kill)
+		if(GCd_at_time > cut_off_time)
 			break // Everything else is newer, skip them
-		queue.Cut(1, 2)
-		var/datum/A
-		A = locate(refID)
-		if (A && A.gc_destroyed == GCd_at_time) // So if something else coincidently gets the same ref, it's not deleted by mistake
-			#ifdef GC_FAILURE_HARD_LOOKUP
-			A.find_references()
-			#endif
-
-			// Something's still referring to the qdel'd object.  Kill it.
-			var/type = A.type
-			testing("GC: -- \ref[A] | [type] was unable to be GC'd and was deleted --")
-			didntgc["[type]"]++
-			
-			HardDelete(A)
-
-			++delslasttick
-			++totaldels
-		else
-			++gcedlasttick
-			++totalgcs
-
-/datum/controller/subsystem/garbage/proc/QueueForQueuing(datum/A)
-	if (istype(A) && A.gc_destroyed == GC_CURRENTLY_BEING_QDELETED)
-		tobequeued += A
-		A.gc_destroyed = GC_QUEUED_FOR_QUEUING
-
-/datum/controller/subsystem/garbage/proc/Queue(datum/A)
-	if (isnull(A) || (!isnull(A.gc_destroyed) && A.gc_destroyed >= 0))
+
+		var/datum/D
+		D = locate(refID)
+
+		if (!D || D.gc_destroyed != GCd_at_time) // So if something else coincidently gets the same ref, it's not deleted by mistake
+			if (level == GC_QUEUE_CHECK)
+				++gcedlasttick
+				++totalgcs
+			continue
+		// Something's still referring to the qdel'd object.
+
+		switch (level)
+			if (GC_QUEUE_CHECK)
+				#ifdef GC_FAILURE_HARD_LOOKUP
+				D.find_references()
+				#endif
+				var/type = D.type
+				testing("GC: -- \ref[D] | [type] was unable to be GC'd and was sent to endo --")
+				didntgc["[type]"]++
+				++delslasttick
+				++totaldels
+			if (GC_QUEUE_ENDODONTICS)
+				EndodonticTherapy(D)
+			if (GC_QUEUE_HARDDELETE)
+				HardDelete(D)
+				continue
+
+		Queue(D, ++level)
+
+	queue -= queued
+	queued = null
+
+
+//a root canal is called a "final restoration" because you basically gut everything out and hope that keeps it from being a pain.
+//	ie: its the final attempt to fix the issue before just removing it.
+/datum/controller/subsystem/garbage/proc/EndodonticTherapy(datum/D)
+	var/static/list/exclude = list("locs", "color", "transform", "parent_type", "vars", "verbs", "type", "gc_destroyed")
+	for (var/V in D.vars - exclude)
+		var/value = D.vars[V]
+		switch(V)
+			if ("ckey")
+				D.vars[V] = null
+				continue
+			if ("tag")
+				D.vars[V] = null
+				continue
+			if ("contents")
+				var/list/L = value
+				if (islist(L))
+					L.Cut()
+				continue
+			if ("loc") //this is where we assume that a turf would never enter the qdel queue. one day this assumption will be wrong.
+				D.vars[V] = null
+				continue
+
+
+		if (islist(value))
+			value -= D
```

Comment:
I was going to null the variable, but in byond modifying certain builtin variables to null can cause oddities and i wanted to avoid adding too many to exempt or adding a second switch tree.

I suppose I could if it passes a byond special list check.